server/src/main/java/org/cloudfoundry/identity/uaa/oauth/provider/client/OAuth2ClientContextFilter.java

```diff
@@ -0,0 +1,153 @@
+
+	/**
+	 * Key in request attributes for the current URI in case it is needed by
+	 * rest client code that needs to send a redirect URI to an authorization
+	 * server.
+	 */
+	public static final String CURRENT_URI = "currentUri";
+
+	private ThrowableAnalyzer throwableAnalyzer = new DefaultThrowableAnalyzer();
+
+	private RedirectStrategy redirectStrategy = new DefaultRedirectStrategy();
+
+	public void afterPropertiesSet() throws Exception {
+		Assert.notNull(redirectStrategy,
+				"A redirect strategy must be supplied.");
+	}
+
+	public void doFilter(ServletRequest servletRequest,
+			ServletResponse servletResponse, FilterChain chain)
+			throws IOException, ServletException {
+		HttpServletRequest request = (HttpServletRequest) servletRequest;
+		HttpServletResponse response = (HttpServletResponse) servletResponse;
+		request.setAttribute(CURRENT_URI, calculateCurrentUri(request));
+
+		try {
+			chain.doFilter(servletRequest, servletResponse);
+		} catch (IOException ex) {
+			throw ex;
+		} catch (Exception ex) {
+			// Try to extract a SpringSecurityException from the stacktrace
+			Throwable[] causeChain = throwableAnalyzer.determineCauseChain(ex);
+			UserRedirectRequiredException redirect = (UserRedirectRequiredException) throwableAnalyzer
+					.getFirstThrowableOfType(
+							UserRedirectRequiredException.class, causeChain);
+			if (redirect != null) {
+				redirectUser(redirect, request, response);
+			} else {
+				if (ex instanceof ServletException servletException) {
+					throw servletException;
+				}
+				if (ex instanceof RuntimeException runtimeException) {
+					throw runtimeException;
+				}
+				throw new NestedServletException("Unhandled exception", ex);
+			}
+		}
+	}
+
+	/**
+	 * Redirect the user according to the specified exception.
+	 * 
+	 * @param e
+	 *            The user redirect exception.
+	 * @param request
+	 *            The request.
+	 * @param response
+	 *            The response.
+	 */
+	protected void redirectUser(UserRedirectRequiredException e,
+			HttpServletRequest request, HttpServletResponse response)
+			throws IOException {
+
+		String redirectUri = e.getRedirectUri();
+		UriComponentsBuilder builder = UriComponentsBuilder
+				.fromHttpUrl(redirectUri);
+		Map<String, String> requestParams = e.getRequestParams();
+		for (Map.Entry<String, String> param : requestParams.entrySet()) {
+			builder.queryParam(param.getKey(), param.getValue());
+		}
+
+		if (e.getStateKey() != null) {
+			builder.queryParam("state", e.getStateKey());
+		}
+
+		this.redirectStrategy.sendRedirect(request, response, builder.build()
+				.encode().toUriString());
+	}
+
+	/**
+	 * Calculate the current URI given the request.
+	 * 
+	 * @param request
+	 *            The request.
+	 * @return The current uri.
+	 */
+	protected String calculateCurrentUri(HttpServletRequest request) {
+		ServletUriComponentsBuilder builder = ServletUriComponentsBuilder
+				.fromRequest(request);
+		// Now work around SPR-10172...
```

Comment:
What is SPR-10172?

Reply:
origininal comment, sometimes these code snippes are used in different spring libraries, see
https://github.com/spring-projects/spring-framework/issues/14805